### PR TITLE
cleanup(node): moves some schematic helper to shared workspace rules …

### PR DESCRIPTION
**Pull request by remshams**
_[Please make sure you have read the submission guidelines before posting an PR](https://github.com/nrwl/nx/blob/master/CONTRIBUTING.md#submit-pr)_

> _Please make sure that your commit message follows our format._

> Example: `fix(nx): must begin with lowercase`

Using the codebase as a really valuable learning opportunity (and writing a node based plugin/package) I moved some things around in the node and express schematic folders.
In case this is useful for you please let me know if I should change something.
Meanwhile I will keep reading... ;)

### CODE REVIEW

**Review by FrozenPandaz — CHANGES_REQUESTED** [review]
Thank you so much for your contribution.

So glad you learned something!

File: packages/workspace/src/utils/rules/workspace.ts [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import { JsonObject } from '@angular-devkit/core';
+import { Rule } from '@angular-devkit/schematics';
+import { updateWorkspace } from '../workspace';
+
+export function setDefaultCollection(collectionName: string): Rule {
```
Comment (FrozenPandaz) [inline packages/workspace/src/utils/rules/workspace.ts]: Please add something in the TSDoc that mentions that this only sets a default collection if it is unset or set to `@nrwl/workspace`

File: packages/workspace/src/utils/rules/workspace.spec.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+} from '@angular-devkit/core/src/workspace/core';
+
+describe('Workspace', () => {
+  const defaultCollectionName = '@nrwl/node';
+  const workspaceJsonFileName = 'workspace.json';
+  const nrwlWorkspaceName = '@nrwl/workspace';
+  let tree: UnitTestTree;
+
+  beforeEach(() => {
+    _test_addWorkspaceFile('workspace.json', WorkspaceFormat.JSON);
```
Comment (FrozenPandaz) [inline packages/workspace/src/utils/rules/workspace.spec.ts]: Why is this necessary?

**Review by FrozenPandaz — CHANGES_REQUESTED** [review]

File: packages/workspace/src/utils/rules/workspace.spec.ts (hunk 2 above)
Comment (remshams) [reply packages/workspace/src/utils/rules/workspace.spec.ts]: If the `workspace.json` is not added as workspace file the `readWorkspace` method (used by `updateWorkspace`) throws an error ("Unable to locate a workspace file for workspace path") as (per default) it only searches for `angular.json` or `.angular.json` in the workspace as workspace file.

As the tests uses the `workspace.json` as test file it is required to register it as a workspace file. If this should be avoided the test would have to be based on a `angular.json` file. I somehow mirrored the approach of the `createEmptyWorkspace` method which also used the `workspace.json` instead of the `angular.json` (probably because the latter requires more work to set up).

An alternative could be to extend these methods to take a kind of loader/reader function which then replaces the `readWorkspace`  method from angular-devkit. However this would also require to make changes to the `updateWorkspace` method (I did not want to touch these "core" methods as I don't know if you want them to be stable/not changed).

File: packages/workspace/src/utils/rules/workspace.ts (hunk 1 above)
Comment (remshams) [reply packages/workspace/src/utils/rules/workspace.ts]: Made an update here, however I was not sure if this what you expected...

**Comment by vsavkin** [comment]
Thank you so much for sending the PR!

I think we can probably get a bit more ambitious with the PR.

- I think most packages in this repo set the default collection property in their init schematic. I think it might makes sense to extend the setDefaultCollection schematic to be something like setWorkspaceDefaults that can set the collection and do other things necessary to setting default. We can then use it in every package.
- Maybe we can combine `removeDependency` with `addDependencies` because the intent is to basically ensure that a particular dependency is added and that is added as a dev dependency.

What do you think?

**Comment by remshams** [comment]
Thank your very much for your feedback and taking the time. As mentioned in my last comment I was not sure if should/could be touching these methods and therefore only made small changes to get some feedback.  
Some thoughts regarding your points:

* When going through the schematics there seem to be two "packages" (angular/react) which do additional work besides setting the `defaultCollection`. As from my understanding in both cases certain properties are set for the relevant schematics like `@nrwl/angular:application` or `@nrwl/react`. So maybe these options could either be provided as a function which will be executed by `setWorkspaceDefaults` or as a dictionary.  
Another option could be to just chain the additional implementation with the `setWorkspaceDefaults` function. The advantage of the former approach would be (in my opinion) that the repetitive `value || {}` checks could be avoided.
* Merging `moveDependency` with `addDependency` (named `makeDevDependency` at the moment for ease of reference) would also work for most cases. Again there are some exceptions (storybook/express) which include adding the dependencies in a separate function e.g. `checkDependenciesInstalled` in the storybook package. However in most of these cases it should be possible to extract the parts covered by `makeDevDependency` from said methods.  
There are several cases where dependencies are added as `dependencies` and not just `devDependencies` for which `makeDevDependency` is not applicable. In these cases either `addDepsToPackageJson` could directly be called (as it is done now) or an additional method could be introduced. Currently this would only be a wrapper but could somehow unify the init procedure chain in the packages or avoid introducing issues. E.g. it seems to to me that the `NodePackageInstallTask` is added twice in the storybook package (in `checkDependenciesInstalled` and `addDepsToPackageJson` tracked by the global `installAdded` variable).  
However I maybe wrong here and there is another reason why it is has to be added in both places..., it just stick out to me while going through the codebase.

**Comment by vsavkin** [comment]
Sorry for the late reply. You made good points.

* Regarding setting the default collection property.  I don't feel very strongly, so I think we can keep what you have. But even then I'd use it in every package that sets the default collection. 
* `removeDependency('@nrwl/express')` is imo a bit problematic because I think it might be misleading. The original helper function was called 'move' because we are moving it from `dependencies` to `devDependencies`. I.e., the intent is not to remove it but to make sure it's set as a dev dependency.  As a side note, the dev/non-dev thing doesn't make much sense nowadays anyways.

**Comment by remshams** [comment]
Thanks again for taking the time and the insights:
  * I will then update the other packages with the function
  * Thankes for the explanation, makes much more sense now.
    As the deps have been deleted from the `dependencies` I thought it was more about removing.
    There is some kind of `addDependencies` method in all packages so could the dependencies just be added there? The `removeDependencies` would then be superfluous.

**Comment by vsavkin** [comment]
@remshams thank you so much for making the changes. I like the solution you came up with. I'm going to merge the PR.

**Comment by remshams** [comment]
@vsavkin  Thanks again for your input/help.
Is there anything else I could do/help e.g. some floor sweeping (no one has time for ;) ) which I could work on in my spare time?
In case not don't bother to comment on the question (you are busy). In that case I will just go through the code.
Thanks
